cle/backends/binja.py

```diff
@@ -0,0 +1,240 @@
+import os
+import logging
+
+from . import Backend, register_backend, Symbol
+from .relocation import Relocation
+from ..errors import CLEError, CLEFileNotFoundError
+from ..address_translator import AT
+import archinfo
+
+l = logging.getLogger("cle.binja")
+
+try:
+    import binaryninja as bn
+except ImportError:
+    l.error("Unable to import binaryninja module")
+
+
+class BinjaSymbol(Symbol):
+    BINJA_FUNC_SYM_TYPES = [bn.SymbolType.ImportedFunctionSymbol,
+                            bn.SymbolType.FunctionSymbol,
+                            bn.SymbolType.ImportAddressSymbol]
+
+    BINJA_DATA_SYM_TYPES = [bn.SymbolType.ImportedDataSymbol,
+                            bn.SymbolType.DataSymbol]
+
+    BINJA_IMPORT_TYPES = [bn.SymbolType.ImportedFunctionSymbol,
+                          bn.SymbolType.ImportAddressSymbol,
+                          bn.SymbolType.ImportedDataSymbol]
+
+    def __init__(self, owner, sym):
+        if sym.type in self.BINJA_FUNC_SYM_TYPES:
+            symtype = Symbol.TYPE_FUNCTION
+        elif sym.type in self.BINJA_DATA_SYM_TYPES:
+            symtype = Symbol.TYPE_OBJECT
+        else:
+            symtype = Symbol.TYPE_OTHER
+        
+        super(BinjaSymbol, self).__init__(owner,
+                                          sym.raw_name,
+                                          AT.from_rva(sym.address, owner).to_rva(),
+                                          owner.bv.address_size,
+                                          symtype)
+
+        if sym.type in self.BINJA_IMPORT_TYPES:
+            self.is_import = True
+
+        # TODO: set is_weak appropriately
+
+class BinjaReloc(Relocation):
+    def __init__(self, owner, symbol, relative_addr):
+        super(BinjaReloc, self).__init__(owner, symbol, relative_addr)
+
+    @property
+    def value(self):
+        return self.relative_addr
+
+
+class BinjaBin(Backend):
+    """
+    Get information from binaries using Binary Ninja. Basing this on idabin.py, but will try to be more complete.
+    TODO: add more features as Binary Ninja's feature set improves
+    """
+    is_default = True # Tell CLE to automatically consider using the BinjaBin backend
+    BINJA_ARCH_MAP = {"aarch64": archinfo.ArchAArch64(endness='Iend_LE'),
+                      "armv7": archinfo.ArchARMEL(endness='Iend_LE'),
+                      "thumb2": archinfo.ArchARMEL(endness='Iend_LE'),
+                      "armv7eb": archinfo.ArchARMEL(endness='Iend_BE'),
+                      "thumb2eb": archinfo.ArchARMEL(endness='Iend_BE'),
+                      "mipsel32": archinfo.ArchMIPS32(endness='Iend_LE'),
+                      "mips32": archinfo.ArchMIPS32(endness='Iend_BE'),
+                      "ppc": archinfo.ArchPPC32(endness="Iend_BE"),
+                      "ppc_le": archinfo.ArchPPC32(endness="Iend_LE"),
+                      "x86": archinfo.ArchX86(),
+                      "x86_64": archinfo.ArchAMD64()}
```

Comment:
In the future it would be nice to make it so that you could just pass any of these strings to `archinfo.arch_from_id()` and have it work automatically

Reply:
Completely agree. I think I ended up doing this because there was a disagreement in default endianness between angr and Binary Ninja when I used that, so I took the more pedantic approach. But yeah, with some tweaking of the binja strings it should be doable.